torch/csrc/autograd/saved_variable_hooks.h

```diff
@@ -0,0 +1,14 @@
+#pragma once
+
+#include <torch/csrc/python_headers.h>
+#include <torch/csrc/WindowsTorchApiMacro.h>
+#include <ATen/ATen.h>
+
+namespace torch { namespace autograd {
+
+struct TORCH_API SavedVariableHooks {
+  virtual PyObject* call_pack_hook(at::Tensor &tensor) = 0;
```

Comment:
So the idea with this interface is to permit C++-only builds to not include any of the `PySavedVariableHooks` logic. 
It's fine because `SavedVariable::register_hooks` is only ever called from Python builds, but this way we don't need to include `PySavedVariableHooks.h` in `SavedVariable.h`

Reply:
Gotcha, maybe name it something like "PySavedVariableHooksInterface" to indicate that it's still python specific. Also you can add this explanation as a comment in that file.

Reply:
I would argue in favor of `SavedVariableHooks`, to honor the trend in the codebase
```c++
// python_hook.h
struct PyFunctionPreHook : public FunctionPreHook {
   // ...
}

// python_hook.h
struct PyFunctionPostHook : public FunctionPostHook {
  // ...
}

// python_engine.h
struct PythonEngine : public Engine {
  // ...
}

// python_function.h
struct PyNode : public Node {
 // ...
}

// python_anomaly_mode.h
struct PyAnomalyMetadata : public AnomalyMetadata {
  // ...
}
```

In particular, `AnomalyMetadata` is employed with the same strategy: it is used in the declaration of the virtual method `Engined::make_anomaly_metadata` whose only implementation is in `python_engine.cpp`. 
For some reason I haven't really dug into, AnomalyMetadata is also a concept that requires Python.

Reply:
But AnomalyMetadata does have a non-python implementation no?
https://github.com/pytorch/pytorch/blob/master/torch/csrc/autograd/anomaly_mode.cpp#L43
Same for all the other examples, all of them can be used in a non-python context

Reply:
In the case of anomaly mode, the feature was first introduced as Python-only: https://github.com/pytorch/pytorch/pull/7677/files

Reply:
I think its okay to call it SavedVariableHooks if we make it Python-agnostic as the comment says
For example have `call_pack_hook` and `call_unpack_hook` take and return Tensors (maybe this was what you planned? I see that you did not import `python_headers.h` in this file.)
Otherwise, if we leave it in its current state, its probably better for the name to reflect the non-python-agnosticness.

Reply:
It's already technically python-agnostic thanks to [this](https://github.com/pytorch/pytorch/blob/master/c10/util/python_stub.h#L4), but I agree: I will rework this to make it easier to subclass if we ever want to implement hooks for C++ frontend.

Reply:
Oh interesting, didn't know about that stub! I guess there's no physical dependence here then - which I guess is "agnostic" some sense.

Reply:
Note from our latest discussion with Victor:
To make this future-proof for an eventual c++ implementation, we most likely dont to make the pack return nothing and the unpack take no input argument.
It is the SavedVariableHooks subclass responsibility to store any data internally. The python version will have a single py::object field for that while a c++ hook will be able to have any type as it sees fit.

Reply:
Yes, that's what ends up being done in #60975. 
I will probably reorganize the stack to make it easier to review.